File: kotlin/motor/src/commonMain/kotlin/com/fintips/motor/Score.kt
package com.fintips.motor

/**
 * Score financeiro de 0 a 100.
 *
 * Cinco dimensões, pesos explícitos, cada ponto com uma fórmula visível. Não
 * há número mágico saído de modelo nenhum — e essa é a razão de o score existir
 * assim: um score opaco não muda comportamento, porque a pessoa não sabe o que
 * fazer com ele. Este diz quantos pontos faltam em qual dimensão e qual
 * alavanca rende mais.
 */

object PesosDoScore {
    const val POUPANCA = 30      // quanto da renda sobra
    const val RESERVA = 25       // meses de despesa cobertos pelo que está guardado
    const val ESTABILIDADE = 15  // previsibilidade do gasto mensal
    const val VAZAMENTOS = 10    // peso do gasto invisível
    const val PLANOS = 20        // aderência às metas cadastradas

    /** Na ordem em que aparecem no painel — e a ordem importa para o harness. */
    val TODOS = listOf(
        "poupanca" to POUPANCA,
        "reserva" to RESERVA,
        "estabilidade" to ESTABILIDADE,
        "vazamentos" to VAZAMENTOS,
        "planos" to PLANOS,
    )
}

/** Alvos de nota cheia. Ficam explícitos porque a alavanca é a distância até eles. */
private const val ALVO_TAXA_POUPANCA = 0.40
private const val ALVO_MESES_RESERVA = 6.0
private const val TETO_VOLATILIDADE = 0.5
private const val TETO_INVISIVEL = 0.15
private const val PENA_POR_MES_NO_VERMELHO = 0.25

data class DimensaoDoScore(val nome: String, val pontos: Double, val maximo: Int) {
    val pct: Int get() = arredondar(pontos / maximo * 100, 0).toInt()
}

data class IndicadoresDoScore(
    val taxaPoupanca: Double,
    val mesesDeReserva: Double,
    val volatilidadeDespesa: Double,
    val gastoInvisivelMes: Double,
    val gastoInvisivelPctDespesa: Double,
    val rendaMediaMes: Double,
    val despesaMediaMes: Double,
)

data class ResultadoDoScore(
    val score: Double,
    val faixa: String,
    val dimensoes: List<DimensaoDoScore>,
    val indicadores: IndicadoresDoScore,
    val proximoPonto: String,
)

/** Aderência de um plano, já avaliada. */
data class AderenciaDePlano(
    val concluido: Boolean,
    val aporteNecessarioMes: Double,
    val capacidadeMensalReal: Double,
)

private fun limitar(x: Double, minimo: Double = 0.0, maximo: Double = 1.0): Double =
    if (x < minimo) minimo else if (x > maximo) maximo else x

private fun faixaDe(score: Double): String = when {
    score >= 85 -> "excelente"
    score >= 70 -> "bom"
    score >= 55 -> "razoavel"
    score >= 40 -> "atencao"
    else -> "critico"
}

/**
 * Calcula o score.
 *
 * Três decisões que parecem detalhe e não são:
 *
 * - **despesa zero não é nota cheia de reserva.** Sem despesa não há quantos
 *   meses cobrir, então a dimensão fica em zero em vez de infinito. Um extrato
 *   sem gasto nenhum é dado incompleto, não saúde financeira.
 * - **mês no vermelho multiplica, não subtrai.** Quatro meses negativos zeram a
 *   estabilidade inteira. Subtrair deixaria um resíduo que sugeriria alguma
 *   previsibilidade onde não há.
 * - **sem plano cadastrado vale meia nota.** Não é prêmio nem castigo: o motor
 *   não sabe se a pessoa não tem metas ou só não as cadastrou, e fingir
 *   qualquer uma das duas seria pior.
 */
fun calcularScore(
    taxaPoupanca: Double,
    volatilidadeDespesa: Double,
    mesesNoVermelho: Int,
    rendaMediaMes: Dinheiro,
    despesaMediaMes: Dinheiro,
    taxasEsegurosMes: Dinheiro,
    microGastosMes: Dinheiro,
    planos: List<AderenciaDePlano>,
    patrimonioLiquido: Dinheiro,
): ResultadoDoScore {
    val despesa = despesaMediaMes.paraDouble()

    // 1. poupança — 40% da renda é o teto da nota
    val pPoupanca = limitar(taxaPoupanca / ALVO_TAXA_POUPANCA) * PesosDoScore.POUPANCA

    // 2. reserva — seis meses de despesa é nota cheia
    val mesesDeReserva = if (despesa > 0) patrimonioLiquido.paraDouble() / despesa else 0.0
    val pReserva = limitar(mesesDeReserva / ALVO_MESES_RESERVA) * PesosDoScore.RESERVA

    // 3. estabilidade — volatilidade de 50% zera; mês no vermelho pune por multiplicação
    var pEstabilidade = limitar(1 - volatilidadeDespesa / TETO_VOLATILIDADE) * PesosDoScore.ESTABILIDADE
    pEstabilidade *= maxOf(0.0, 1 - PENA_POR_MES_NO_VERMELHO * mesesNoVermelho)

    // 4. vazamentos — invisível acima de 15% da despesa zera a dimensão
    val invisivelMes = taxasEsegurosMes + microGastosMes
    val pctInvisivel = if (despesa > 0) invisivelMes.paraDouble() / despesa else 0.0
    val pVazamentos = limitar(1 - pctInvisivel / TETO_INVISIVEL) * PesosDoScore.VAZAMENTOS

    // 5. planos — média de aderência; sem plano cadastrado, metade da nota
    val pPlanos = if (planos.isNotEmpty()) {
        val aderencias = planos.map { p ->
            when {
                p.concluido -> 1.0
                p.aporteNecessarioMes <= 0 -> 0.5
                else -> limitar(p.capacidadeMensalReal / p.aporteNecessarioMes)
            }
        }
        aderencias.sum() / aderencias.size * PesosDoScore.PLANOS
    } else {
        PesosDoScore.PLANOS * 0.5
    }

    val partes = linkedMapOf(
        "poupanca" to arredondar(pPoupanca, 1),
        "reserva" to arredondar(pReserva, 1),
        "estabilidade" to arredondar(pEstabilidade, 1),
        "vazamentos" to arredondar(pVazamentos, 1),
        "planos" to arredondar(pPlanos, 1),
    )
    val total = arredondar(partes.values.sum(), 1)

    return ResultadoDoScore(
        score = total,
        faixa = faixaDe(total),
        dimensoes = PesosDoScore.TODOS.map { (nome, peso) ->
            DimensaoDoScore(nome, partes[nome]!!, peso)
        },
        indicadores = IndicadoresDoScore(
            taxaPoupanca = arredondar(taxaPoupanca * 100, 1),
            mesesDeReserva = arredondar(mesesDeReserva, 1),
            volatilidadeDespesa = arredondar(volatilidadeDespesa, 3),
            gastoInvisivelMes = invisivelMes.paraDouble(),
            gastoInvisivelPctDespesa = arredondar(pctInvisivel * 100, 1),
            rendaMediaMes = rendaMediaMes.paraDouble(),
            despesaMediaMes = despesa,
        ),
        proximoPonto = proximaAlavanca(partes),
    )
}

/**
 * A dimensão com maior ganho marginal — onde o esforço rende mais ponto.
 *
 * Empate fica com a primeira na ordem dos pesos, como no Python, onde o `max`
 * sobre o dicionário devolve a primeira chave de valor máximo. Parece
 * arbitrário e é: o importante é ser **o mesmo** arbitrário dos dois lados.
 */
private fun proximaAlavanca(partes: Map<String, Double>): String {
    val dicas = mapOf(
        "poupanca" to "aumentar a sobra mensal: o maior ganho de score está em guardar mais do que entra",
        "reserva" to "engordar a reserva até cobrir 6 meses de despesa",
        "estabilidade" to "reduzir a variação do gasto entre meses — orçamento por categoria",
        "vazamentos" to "cortar taxas, seguros e micro-gastos recorrentes",
        "planos" to "ajustar prazo ou aporte dos planos para caberem na capacidade real",
    )
    var alvo = PesosDoScore.TODOS.first().first
    var maiorLacuna = Double.NEGATIVE_INFINITY
    for ((nome, peso) in PesosDoScore.TODOS) {
        val lacuna = peso - (partes[nome] ?: 0.0)
        if (lacuna > maiorLacuna) {
            maiorLacuna = lacuna
            alvo = nome
        }
    }
    return dicas[alvo]!!
}

File: fintips/score.py
"""Score financeiro 0–100.

Cinco dimensões, pesos explícitos, tudo rastreável: cada ponto tem uma
fórmula visível. Nada de número mágico saído do LLM.
"""

from __future__ import annotations

from decimal import Decimal

Z = Decimal("0")

WEIGHTS = {
    "poupanca": 30,     # quanto da renda sobra
    "reserva": 25,      # meses de despesa cobertos pelo que está guardado
    "estabilidade": 15, # previsibilidade do gasto mensal
    "vazamentos": 10,   # peso dos gastos invisíveis
    "planos": 20,       # aderência às metas cadastradas
}


def _clamp(x: float, lo: float = 0.0, hi: float = 1.0) -> float:
    return max(lo, min(hi, x))


def _band(score: float) -> str:
    if score >= 85:
        return "excelente"
    if score >= 70:
        return "bom"
    if score >= 55:
        return "razoavel"
    if score >= 40:
        return "atencao"
    return "critico"


def compute(
    base: dict,
    invisible: dict,
    plan_results: list[dict],
    *,
    patrimonio_liquido: Decimal | float = 0,
) -> dict:
    patrimonio = Decimal(str(patrimonio_liquido))
    despesa = Decimal(str(base.get("despesa_media_mes", 0) or 0))
    renda = Decimal(str(base.get("renda_media_mes", 0) or 0))

    # 1. Taxa de poupança — 20% já é saudável, 40%+ é teto da nota.
    taxa = float(base.get("taxa_poupanca_media", 0) or 0)
    p_poupanca = _clamp(taxa / 0.40) * WEIGHTS["poupanca"]

    # 2. Reserva de emergência — 6 meses de despesa = nota cheia.
    meses_reserva = float(patrimonio / despesa) if despesa > 0 else 0.0
    p_reserva = _clamp(meses_reserva / 6.0) * WEIGHTS["reserva"]

    # 3. Estabilidade — volatilidade de 0 é ótima, 50%+ zera; mês no vermelho pune.
    vol = float(base.get("volatilidade_despesa", 0) or 0)
    p_estab = _clamp(1 - vol / 0.5) * WEIGHTS["estabilidade"]
    vermelhos = int(base.get("meses_no_vermelho", 0) or 0)
    p_estab *= max(0.0, 1 - 0.25 * vermelhos)

    # 4. Vazamentos — invisível acima de 15% da despesa zera a dimensão.
    invis_mes = (
        Decimal(str(invisible.get("taxas_e_seguros", {}).get("por_mes", 0)))
        + Decimal(str(invisible.get("micro_gastos", {}).get("por_mes", 0)))
    )
    pct_invis = float(invis_mes / despesa) if despesa > 0 else 0.0
    p_vaz = _clamp(1 - pct_invis / 0.15) * WEIGHTS["vazamentos"]

    # 5. Planos — média de aderência; sem plano cadastrado, metade da nota.
    if plan_results:
        aderencias = []
        for r in plan_results:
            nec = r.get("aporte_necessario_mes") or 0
            cap = r.get("capacidade_mensal_real") or 0
            if r.get("status") == "concluido":
                aderencias.append(1.0)
            elif nec <= 0:
                aderencias.append(0.5)
            else:
                aderencias.append(_clamp(cap / nec))
        p_planos = (sum(aderencias) / len(aderencias)) * WEIGHTS["planos"]
    else:
        p_planos = WEIGHTS["planos"] * 0.5

    partes = {
        "poupanca": round(p_poupanca, 1),
        "reserva": round(p_reserva, 1),
        "estabilidade": round(p_estab, 1),
        "vazamentos": round(p_vaz, 1),
        "planos": round(p_planos, 1),
    }
    total = round(sum(partes.values()), 1)

    return {
        "score": total,
        "faixa": _band(total),
        "dimensoes": [
            {
                "nome": k,
                "pontos": partes[k],
                "maximo": WEIGHTS[k],
                "pct": round(partes[k] / WEIGHTS[k] * 100),
            }
            for k in WEIGHTS
        ],
        "indicadores": {
            "taxa_poupanca": round(taxa * 100, 1),
            "meses_de_reserva": round(meses_reserva, 1),
            "volatilidade_despesa": round(vol, 3),
            "gasto_invisivel_mes": float(invis_mes),
            "gasto_invisivel_pct_despesa": round(pct_invis * 100, 1),
            "renda_media_mes": float(renda),
            "despesa_media_mes": float(despesa),
        },
        "proximo_ponto": _next_lever(partes),
    }


def _next_lever(partes: dict[str, float]) -> str:
    gaps = {k: WEIGHTS[k] - v for k, v in partes.items()}
    alvo = max(gaps, key=gaps.get)
    dicas = {
        "poupanca": "aumentar a sobra mensal: o maior ganho de score está em guardar mais do que entra",
        "reserva": "engordar a reserva até cobrir 6 meses de despesa",
        "estabilidade": "reduzir a variação do gasto entre meses — orçamento por categoria",
        "vazamentos": "cortar taxas, seguros e micro-gastos recorrentes",
        "planos": "ajustar prazo ou aporte dos planos para caberem na capacidade real",
    }
    return dicas[alvo]
